### PR TITLE
[MPM] particle generator fix

### PR DESCRIPTION
**Pull request by peterjwilson**
Hi MPM team, this is a proposed fix for the particle generator. During particle generation, it was previously assumed that 'n' MPs generated from an element with mass 'M' and volume 'V' had equal volumes and mass according to 'm[i] = M / n' and 'v[i] = V / n'. This is only true for perfectly undistorted elements. Now integration point volumes are used to proportion out the total element mass and volume in the correct manner.

There's two questions I'd like to get your feedback on please (@tteschemacher , @bodhinandach , @VeronikaSinger , @antonialarese ):

1.  ~~This change will probably break alot of tests. I'm happy to update them all, but before I do it would be great to get the go-ahead from all you guys that this is the correct approach and method.~~ _Update: All the tests seem to pass fine, they must have all used undistorted elements_

2.  I have left in the custom cases of 2d tri 16 and 33 particles per element. Since integration weights can't be generated via integration points() for these I've just added a warning that the geometry needs to be perfectly undistorted. Another option would be to remove the 16 and 33 particles per element options entirely. I'm easy either way.

### CODE REVIEW

**Comment by philbucher** [comment]
this utility is substantial for the MPM-stuffs, I highly recommend adding many tests (esp for corner-cases like this) to be save for future refactorings

File: applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.h [hunk 1]
```diff
@@ -118,30 +124,38 @@ namespace MPMParticleGeneratorUtility
 
                     const Geometry< Node < 3 > >& r_geometry = i->GetGeometry(); // current element's geometry
                     const GeometryData::KratosGeometryType geo_type = r_geometry.GetGeometryType();
-                    Matrix shape_functions_values = r_geometry.ShapeFunctionsValues(GeometryData::GI_GAUSS_2);
+                    auto int_method = GeometryData::GI_GAUSS_2;
+                    Matrix shape_functions_values;
+                    bool is_equal_int_volumes = false;
                     if (geo_type == GeometryData::Kratos_Tetrahedra3D4 || geo_type == GeometryData::Kratos_Triangle2D3)
```
Comment (tteschemacher) [inline applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.h]: could you wrap the following within a separate function?

**Comment by tteschemacher** [comment]
I fully agree. It has already a certain set of tests. Some even for this function only. But it could be more.

I suppose the tests would fail as soon as the accuracy of the result comparisons would be increased significantly.

Btw. this function will be a modeler in close future.

Thank you @PeterJosephWilson for this change!

**Review by tteschemacher — DISMISSED** [review]
Thanks, we can discuss the other proposed issue within another PR.

**Comment by peterjwilson** [comment]
Added tests @philbucher and externalized the geometry switch @tteschemacher

**Review by tteschemacher — APPROVED** [review]
It makes the generator much more handsome. The test looks logical and tests additional features.

**Pull request merged**